probe_src/python/probe_py/manual/persistent_provenance.py

```diff
@@ -0,0 +1,98 @@
+from __future__ import annotations
+import typing
+import json
+import datetime
+import dataclasses
+import pathlib
+import xdg_base_dirs
+
+
+@dataclasses.dataclass(frozen=True)
+class Inode:
+    host: int
+    device_major: int
+    device_minor: int
+    inode: int
+
+    def str_id(self) -> str:
+        return f"{self.host:012x}-{self.device_major:04x}-{self.device_minor:04x}-{self.inode:016x}"
+
+
+@dataclasses.dataclass(frozen=True)
+class InodeVersion:
+    inode: Inode
+
+    # Usually, different versions of the inode will have different mtimes
+    # But not always.
+    # The file could be modified multiple times within the granularity of the system's mtime clock or the user could change the mtime
+    mtime: int
+
+    # Size is included to give a simple "pseudo-hash" of the content
+    # Different size implies different content, but not necessarily the converse: same size does not necessarily imply same content.
+    # Size serves as an inexpensive check of whether the previous attributes uniquely identify the same contents
+    size: int
+
+    # TODO: Handle the case where the user manually changes the mtime without changing the size.
+    # We have no way to tell versions if the user resets the mtime of a new version to the mtime of a previous version.
+    # Possible solutions:
+    # - Wrap, intercept, and warn *utime* lib calls,
+    # - Use an xattr, if the underlying FS supports it.
+    #   Our xattr would be "true mtime" that we would maintain and can't be changed by normal tools.
+    # - Cry.
+
+    def str_id(self) -> str:
+        return f"{self.inode.str_id()}-{self.mtime:016x}-{self.size:016x}"
+
+
+@dataclasses.dataclass(frozen=True)
+class InodeMetadataVersion:
+    inode_version: InodeVersion
+    stat_results: bytes
```

Comment:
Probably not a big deal  (this'll be a moot point when this is moved to rust anyway), but we should have a dataclass to represent a stat in the `probe_py.generated`.